In [50]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>


In [51]:
tests = [
    {
        "name": "Example 1",
        "fr": 1,
        "to": 3,
        "symbol": "a",
        "password": "abcde",
        "expected": True,
    },
    {
        "name": "Example 2",
        "fr": 1,
        "to": 3,
        "symbol": "b",
        "password": "cdefg",
        "expected": False,
    },
    {
        "name": "Example 3",
        "fr": 2,
        "to": 9,
        "symbol": "c",
        "password": "ccccccccc",
        "expected": True,
    },
]


def valid(fr: int, to: int, symbol: chr, password: str) -> bool:
    return fr <= password.count(symbol) <= to


run_tests_params(valid, tests)


example = """
    1-3 a: abcde
    1-3 b: cdefg
    2-9 c: ccccccccc
"""


def parse_line(s: str) -> tuple[int, int, chr, str]:
    fr_to, symbol, password = s.strip().split()

    fr, to = (int(i) for i in fr_to.split("-"))
    symbol = symbol[:-1]
    return fr, to, symbol, password


def parse(s):
    return (parse_line(l) for l in s.strip().splitlines())


def count_valid_passwords(s: str) -> int:
    return sum(
        1 for fr, to, symbol, password in parse(s) if valid(fr, to, symbol, password)
    )


assert count_valid_passwords(example) == 2


Test Example 1 passed, for valid.
Test Example 2 passed, for valid.
Test Example 3 passed, for valid.
Success


In [52]:
with open("../input/day2.txt") as f:
    puzzle = f.read()

print(f"Part I: {count_valid_passwords(puzzle)}")

Part I: 393


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>393</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>While it appears you validated the passwords correctly, they don't seem to be what the Official Toboggan Corporate Authentication System is expecting.</p>
<p>The shopkeeper suddenly realizes that he just accidentally explained the password policy rules from his old job at the sled rental place down the street! The Official Toboggan Corporate Policy actually works a little differently.</p>
<p>Each policy actually describes two <em>positions in the password</em>, where <code>1</code> means the first character, <code>2</code> means the second character, and so on. (Be careful; Toboggan Corporate Policies have no concept of "index zero"!) <em>Exactly one of these positions</em> must contain the given letter. Other occurrences of the letter are irrelevant for the purposes of policy enforcement.</p>
<p>Given the same example list from above:</p>
<ul>
<li><code>1-3 a: <em>a</em>b<em>c</em>de</code> is <em>valid</em>: position <code>1</code> contains <code>a</code> and position <code>3</code> does not.</li>
<li><code>1-3 b: <em>c</em>d<em>e</em>fg</code> is <em>invalid</em>: neither position <code>1</code> nor position <code>3</code> contains <code>b</code>.</li>
<li><code>2-9 c: c<em>c</em>cccccc<em>c</em></code> is <em>invalid</em>: both position <code>2</code> and position <code>9</code> contain <code>c</code>.</li>
</ul>
<p><em>How many passwords are valid</em> according to the new interpretation of the policies?</p>
</article>

</main>


In [53]:
tests = [
    {
        "name": "Example 1",
        "fr": 1,
        "to": 3,
        "symbol": "a",
        "password": "abcde",
        "expected": True,
    },
    {
        "name": "Example 2",
        "fr": 1,
        "to": 3,
        "symbol": "b",
        "password": "cdefg",
        "expected": False,
    },
    {
        "name": "Example 3",
        "fr": 2,
        "to": 9,
        "symbol": "c",
        "password": "ccccccccc",
        "expected": False,
    },
]


def valid(fr: int, to: int, symbol: chr, password: str) -> bool:
    return (password[fr - 1] == symbol) ^ (password[to - 1] == symbol)


run_tests_params(valid, tests)


example = """
    1-3 a: abcde
    1-3 b: cdefg
    2-9 c: ccccccccc
"""


assert count_valid_passwords(example) == 1


Test Example 1 passed, for valid.
Test Example 2 passed, for valid.
Test Example 3 passed, for valid.
Success


In [54]:
print(f"Part II: {count_valid_passwords(puzzle)}")

Part II: 690


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>690</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
